Lib/test/test_asyncio/test_subprocess.py

```diff
@@ -672,12 +672,13 @@ def setUp(self):
             policy.set_child_watcher(watcher)
 
         def tearDown(self):
-            super().tearDown()
             policy = asyncio.get_event_loop_policy()
             watcher = policy.get_child_watcher()
             policy.set_child_watcher(None)
             watcher.attach_loop(None)
             watcher.close()
+            # Since setUp() does super().setUp() first, do tearDown() last.
+            super().tearDown()
```

Comment:
I think this change causes the CI to fail.

Reply:
Calling `super().tearDown()` last here should not be causing dangling threads, so I think this could be pointing to a more fundamental issue.

In a previous PR, I added an internal `_join_threads()` method to ThreadedChildWatcher, which specifically cleans up all non-daemon threads. This was there to ensure all threads are joined when the watcher is closed. However, in `add_child_handler` the `_do_waitpid()` threads that are created are set as daemon, meaning they aren't joined in `_join_threads` since it checks if they're not daemon.

Based on the error message in Travis, it definitely seems to be those waitpid threads created in ThreadedChildWatcher causing the hangs:

```
0:11:41 load avg: 3.89 [375/424/1] test_asyncio failed (env changed) (2 min 3 sec) -- running: test_multiprocessing_spawn (1 min 2 sec)
Warning -- threading_cleanup() failed to cleanup 1 threads (count: 1, dangling: 2)
Warning -- Dangling thread: <_MainThread(MainThread, started 140164662380352)>
Warning -- Dangling thread: <Thread(waitpid-0, started daemon 140164437370624)>
Warning -- threading_cleanup() failed to cleanup 1 threads (count: 1, dangling: 2)
Warning -- Dangling thread: <Thread(waitpid-0, started daemon 140164437370624)>
Warning -- Dangling thread: <_MainThread(MainThread, started 140164662380352)>
Warning -- threading_cleanup() failed to cleanup 1 threads (count: 1, dangling: 2)
Warning -- Dangling thread: <_MainThread(MainThread, started 140164662380352)>
Warning -- Dangling thread: <Thread(waitpid-0, started daemon 140164437370624)>
Warning -- threading_cleanup() failed to cleanup 1 threads (count: 1, dangling: 2)
Warning -- Dangling thread: <_MainThread(MainThread, started 140164662380352)>
Warning -- Dangling thread: <Thread(waitpid-0, started daemon 140164437370624)>
```

As a result, I think a viable fix might be just removing the and `not not thread.daemon` check in `ThreadedChildWatcher._join_threads()`, to have it also clean up the waitpid threads when the watcher is closed. Based on my local testing, this seems to resolve the issue, but `test-with-buildbots` should probably be used to make sure it doesn't cause other side effects.

Reply:
I don't recall exactly why I made this particular `tearDown()` change for the purposes of this PR. (There was probably some failure I was trying to see if I could get by quickly.) In any case, it should be incidental to the `MultiLoopChildWatcher` hang. So if any changes are needed for the asyncio tests related to `tearDown()`, it should be done in a separate PR IMO independent of the `MultiLoopChildWatcher` hang issue. In other words, this eventual PR shouldn't contain this change.

Reply:
@aeros Just to clarify what I said above, if you feel your `ThreadedChildWatcher` change would help the test harness in general, I would recommend making that change in a separate PR to reduce the number of moving parts in the changes needed to address the `MultiLoopChildWatcher` hang.

Reply:
Alright, I'll look into opening a separate PR to address it. However, if this PR is continued upon (either by yourself or me), I'd suggest reverting the `super().tearDown()` location change made.

Reply:
Right. I guess there are a few steps to be done. First, you can do your `ThreadedChildWatcher` change in a separate PR, and `super().tearDown()` can be reverted in this PR. If however, after resuming work on this PR, it looks like a change to the teardown logic might be needed for some reason, I would still suggest that the `tearDown()` change also be done in a separate PR (since it affects many of the watcher tests and not just `MultiLoopChildWatcher`, and so would be good to isolate that change).